### ISSUE TITLE
Spider Exceptions; catching or re-raising

### ISSUE THREAD

**Issue by mohmad-null** [scrapy/scrapy#6436]
There seems to be very little documentation on catching exceptions with scrapy. But before I open issue(s?) around that, I wanted to check.

My code has a `raise MemoryError()` which correctly triggers. Unfortunately scrapy (via `handle_spider_error()`) simply logs it and carries on, which is absolutely not desired.
Test code:

```
def errback_httpbin(self, failure):
	raise MemoryError('broken stuff')
```

1) As far as I can see, there's no documentation around `handle_spider_error()`. Question: Does scrapy have any way of catching/re-raising specified exceptions? Or is is just "every exception except CloseSpider is caught and logged"?

2) https://github.com/scrapy/scrapy/issues/1015 - is this broken again? The middleware I created (copy/pasted from that ticket) doesn't get triggered despite appearing in the logs as being enabled, and its `def _init_(self): print(111)` runs. Errors only go to `handle_spider_error()`.

```
2024-07-13 12:35:10 [scrapy.middleware] INFO: Enabled spider middlewares:
['scrapy.spidermiddlewares.httperror.HttpErrorMiddleware',
 'scrapy.spidermiddlewares.offsite.OffsiteMiddleware',
 'src.spiders.middlewares.ManyExceptionsMiddleware',
 'scrapy.spidermiddlewares.urllength.UrlLengthMiddleware',
 'scrapy.spidermiddlewares.depth.DepthMiddleware']
```

3) https://webscraping.ai/faq/scrapy/how-do-i-handle-errors-and-exceptions-in-scrapy that page says you can override the `handle_spider_error()`, but it doesn't work for me; `from_crawler()` gets executed, but then when the exception is hit, it just uses the original `handle_spider_error()`. This code:
```
from scrapy import signals

class MySpider(scrapy.Spider):
    name = 'myspider'

    @classmethod
    def from_crawler(cls, crawler, *args, **kwargs):
        spider = super().from_crawler(crawler, *args, **kwargs)
        crawler.signals.connect(spider.handle_spider_error, signal=signals.spider_error)
        return spider

    def handle_spider_error(self, failure, response, spider):
        self.logger.error(f"A spider error occurred: {failure}")
```

There must be a better way to catch and handle specific exceptions than this. Am I missing something?

**Comment by wRAR** [comment]
AFAIK `handle_spider_error()` works like that by design, which is "an exception in the callback must not break the spider". If the user wants to catch exceptions raised from a callback they must use a spider middleware with `process_spider_exception()`.

> process_spider_exception not called with exception from spider #1015 - is this broken again? 

No.

> https://webscraping.ai/faq/scrapy/how-do-i-handle-errors-and-exceptions-in-scrapy that page says you can override the handle_spider_error()

It doesn't say that. It says you can create a function and name it `handle_spider_error()`, which is unrelated to `Scraper.handle_spider_error()`.

**Comment by wRAR** [comment]
And yeah, making a handler for the `spider_error` signal indeed looks like a good alternative to making a spider middleware when it's a better option, so the article is right in that (despite being typical AI-generated crap otherwise).

**Comment by mohmad-null** [comment]
Thanks for the clarifications. I can confirm that `process_spider_exception()` simply isn't working for me.

Running in debug mode, I can confirm that `SpiderMiddlewareManager._add_middleware()` picks up my middleware, however it absolutely doesn't get called ever during an exception. I tried following the code through to see where it was missing the call to the middleware, but it's a rats nest of callback stuff I don't understand.

So I added a breakpoint to every single use of `process_spider_exception` in the scrapy codebase (all are in `spidermw.py`, except one in `httperror.py`) - not a single one gets called other than the two lines in `SpiderMiddlewareManager._add_middleware()`. The exception only goes to `handle_spider_error()`. This is true whether my own middleware is declared or not.

`SpiderMiddlewareManager._process_spider_exception()` is never called, nor is `SpiderMiddlewareManager._process_spider_output()`, nor `SpiderMiddlewareManager.scrape_response.process_spider_exception()`.

I don't see anything in https://docs.scrapy.org/en/latest/topics/spider-middleware.html#scrapy.spidermiddlewares.SpiderMiddleware.process_spider_exception that suggests there are any caveats or exceptions (ha!) to using this method.
